File: typescript/src/ancilis/dependencies/detector.ts
/** Lockfile detection and dependency parsing for npm ecosystems. */

import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYaml } from "yaml";
import type { Dependency, DetectionResult } from "./types.js";

// Priority order: pnpm > yarn > npm
const MANIFEST_CANDIDATES = [
  "pnpm-lock.yaml",
  "yarn.lock",
  "package-lock.json",
] as const;

type ManifestFormat = "pnpm-lock.yaml" | "yarn.lock" | "package-lock.json";

// ---------------------------------------------------------------------------
// package-lock.json
// ---------------------------------------------------------------------------

interface PackageLockV2 {
  lockfileVersion?: number;
  packages?: Record<string, { version?: string; dev?: boolean }>;
  dependencies?: Record<string, { version: string }>;
}

function parsePackageLock(content: string): Dependency[] {
  let data: PackageLockV2;
  try {
    data = JSON.parse(content) as PackageLockV2;
  } catch {
    return [];
  }

  const deps: Dependency[] = [];

  if (data.lockfileVersion && data.lockfileVersion >= 2 && data.packages) {
    for (const [path, info] of Object.entries(data.packages)) {
      // Skip the root package entry (empty string key or workspace root)
      if (!path || !path.startsWith("node_modules/")) continue;
      if (!info.version) continue;
      // Extract package name from path like "node_modules/foo" or "node_modules/@scope/foo"
      const name = path.replace(/^node_modules\//, "");
      // Skip nested node_modules (e.g. "node_modules/foo/node_modules/bar")
      if (name.includes("node_modules")) continue;
      deps.push({ name, version: info.version, ecosystem: "npm" });
    }
  } else if (data.dependencies) {
    // v1 format
    for (const [name, info] of Object.entries(data.dependencies)) {
      if (info.version) {
        deps.push({ name, version: info.version, ecosystem: "npm" });
      }
    }
  }

  return deps;
}

// ---------------------------------------------------------------------------
// yarn.lock  (classic v1 and Berry v2+)
// ---------------------------------------------------------------------------

function parseYarnLock(content: string): Dependency[] {
  const deps: Dependency[] = [];
  const lines = content.split("\n");
  let currentNames: string[] = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]!;

    // Skip comments and empty lines
    if (line.startsWith("#") || line.trim() === "") {
      currentNames = [];
      continue;
    }

    // Berry YAML-ish header: `"foo@npm:^1.0.0":` or classic: `foo@^1.0.0:`
    // Both end with `:` and don't start with spaces
    if (!line.startsWith(" ") && line.endsWith(":")) {
      const specifiers = line.slice(0, -1).split(",").map((s) => s.trim());
      currentNames = specifiers
        .map((s) => {
          // Remove surrounding quotes
          const unquoted = s.replace(/^"|"$/g, "");
          // Berry npm protocol: "foo@npm:^1.0.0"
          const berryMatch = unquoted.match(/^(@?[^@]+)@npm:/);
          if (berryMatch) return berryMatch[1] ?? "";
          // Classic: foo@^1.0.0 or @scope/foo@^1.0.0
          const classicMatch = unquoted.match(/^(@?[^@]+)@/);
          return classicMatch ? (classicMatch[1] ?? "") : "";
        })
        .filter(Boolean);
      continue;
    }

    // Version line: `  version "1.2.3"` or `  version: 1.2.3`
    const versionMatch = line.match(/^\s+version[:\s]+"?([^\s"]+)"?/);
    if (versionMatch && currentNames.length > 0) {
      const version = versionMatch[1]!;
      for (const name of currentNames) {
        deps.push({ name, version, ecosystem: "npm" });
      }
      currentNames = [];
      continue;
    }
  }

  // Deduplicate by name (keep first seen)
  const seen = new Set<string>();
  return deps.filter((d) => {
    if (seen.has(d.name)) return false;
    seen.add(d.name);
    return true;
  });
}

// ---------------------------------------------------------------------------
// pnpm-lock.yaml
// ---------------------------------------------------------------------------

function parsePnpmLock(content: string): Dependency[] {
  let data: Record<string, unknown>;
  try {
    data = parseYaml(content) as Record<string, unknown>;
  } catch {
    return [];
  }

  const deps: Dependency[] = [];

  // pnpm v6/v7/v8: packages object with keys like "/lodash@4.17.21" or "/lodash/4.17.21"
  // pnpm v9: keys like "lodash@4.17.21"
  const packages = data["packages"] as Record<string, unknown> | undefined;
  if (!packages) return [];

  for (const key of Object.keys(packages)) {
    // Strip leading slash
    const stripped = key.startsWith("/") ? key.slice(1) : key;

    // pnpm v9 format: "name@version" or "@scope/name@version"
    // pnpm v6 format: "name/version" or "@scope/name/version"
    let name: string;
    let version: string;

    if (stripped.includes("@") && !stripped.startsWith("@")) {
      // Simple scoped: "lodash@4.17.21"
      const atIdx = stripped.lastIndexOf("@");
      name = stripped.slice(0, atIdx);
      version = stripped.slice(atIdx + 1);
    } else if (stripped.startsWith("@")) {
      // Scoped package: "@scope/pkg@1.0.0"
      // Find the @version part after the package name
      const secondAt = stripped.indexOf("@", 1);
      if (secondAt === -1) {
        // pnpm v6: "@scope/pkg/1.0.0"
        const slashIdx = stripped.lastIndexOf("/");
        name = stripped.slice(0, slashIdx);
        version = stripped.slice(slashIdx + 1);
      } else {
        name = stripped.slice(0, secondAt);
        version = stripped.slice(secondAt + 1);
      }
    } else {
      // pnpm v6 non-scoped: "name/version"
      const slashIdx = stripped.lastIndexOf("/");
      if (slashIdx === -1) continue;
      name = stripped.slice(0, slashIdx);
      version = stripped.slice(slashIdx + 1);
    }

    // Strip any trailing peer specifiers like "_foo@1.0.0"
    version = version.split("_")[0]!;

    if (name && version) {
      deps.push({ name, version, ecosystem: "npm" });
    }
  }

  return deps;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export function detectDependencies(projectDir: string): DetectionResult | null {
  for (const filename of MANIFEST_CANDIDATES) {
    const fullPath = join(projectDir, filename);
    if (!existsSync(fullPath)) continue;

    const content = readFileSync(fullPath, "utf-8");
    let dependencies: Dependency[];

    switch (filename as ManifestFormat) {
      case "pnpm-lock.yaml":
        dependencies = parsePnpmLock(content);
        break;
      case "yarn.lock":
        dependencies = parseYarnLock(content);
        break;
      case "package-lock.json":
        dependencies = parsePackageLock(content);
        break;
    }

    return { manifestPath: fullPath, manifestFormat: filename as ManifestFormat, dependencies };
  }

  return null;
}

File: python/src/ancilis/dependencies/detector.py
"""Lockfile detection and dependency parsing for Python projects.

Priority order: poetry.lock > Pipfile.lock > requirements.txt > pyproject.toml
Returns the first manifest found (highest-priority match).
"""

from __future__ import annotations

import json
import re
import sys
import warnings
from dataclasses import dataclass, field
from pathlib import Path

if sys.version_info >= (3, 11):
    import tomllib
else:
    try:
        import tomllib
    except ImportError:
        tomllib = None


@dataclass
class Dependency:
    name: str
    version: str
    ecosystem: str = "PyPI"


@dataclass
class DetectionResult:
    manifest_path: str
    manifest_format: str  # "poetry.lock" | "Pipfile.lock" | "requirements.txt" | "pyproject.toml"
    dependencies: list[Dependency] = field(default_factory=list)


# ---------------------------------------------------------------------------
# requirements.txt
# ---------------------------------------------------------------------------

_REQUIREMENTS_PINNED = re.compile(
    r"^\s*([A-Za-z0-9_.\-]+)(?:\[.*?\])?==([^\s;#]+)", re.IGNORECASE
)


def _parse_requirements_txt(path: Path) -> list[Dependency]:
    deps: list[Dependency] = []
    try:
        lines = path.read_text(encoding="utf-8").splitlines()
    except OSError:
        return deps
    has_unpinned = False
    for line in lines:
        line = line.strip()
        if not line or line.startswith("#") or line.startswith("-"):
            continue
        m = _REQUIREMENTS_PINNED.match(line)
        if m:
            deps.append(Dependency(name=m.group(1), version=m.group(2)))
        else:
            # Check if it's a real specifier (not a URL or option)
            if re.match(r"^[A-Za-z0-9_.\-]+", line):
                has_unpinned = True
    if has_unpinned:
        warnings.warn(
            f"{path}: some dependencies lack pinned versions — OSV results may be incomplete",
            UserWarning,
            stacklevel=2,
        )
    return deps


# ---------------------------------------------------------------------------
# pyproject.toml
# ---------------------------------------------------------------------------


def _normalise_pep508(spec: str) -> tuple[str, str | None]:
    spec = re.sub(r"\[.*?\]", "", spec).split(";")[0].strip()
    m = re.match(r"^([A-Za-z0-9_.\-]+)==([^\s,]+)", spec)
    if m:
        return m.group(1), m.group(2)
    m = re.match(r"^([A-Za-z0-9_.\-]+)", spec)
    if m:
        return m.group(1), None
    return spec, None


def _parse_pyproject_toml(path: Path) -> list[Dependency]:
    if tomllib is None:
        return []
    try:
        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return []
    raw_deps: list[str] = []
    # [project.dependencies] (PEP 621)
    raw_deps.extend(data.get("project", {}).get("dependencies", []))
    # [tool.poetry.dependencies]
    for name, val in data.get("tool", {}).get("poetry", {}).get("dependencies", {}).items():
        if name.lower() == "python":
            continue
        if isinstance(val, str):
            raw_deps.append(f"{name}{val}" if val.startswith(("=", "^", "~", ">", "<")) else name)
        elif isinstance(val, dict) and "version" in val:
            raw_deps.append(f"{name}{val['version']}")
    deps: list[Dependency] = []
    for spec in raw_deps:
        name, version = _normalise_pep508(str(spec))
        if version is not None:
            deps.append(Dependency(name=name, version=version))
    return deps


# ---------------------------------------------------------------------------
# Pipfile.lock
# ---------------------------------------------------------------------------


def _parse_pipfile_lock(path: Path) -> list[Dependency]:
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return []
    deps: list[Dependency] = []
    for section in ("default", "develop"):
        for pkg, info in data.get(section, {}).items():
            if isinstance(info, dict):
                version_raw = info.get("version", "")
                m = re.match(r"==(.+)", version_raw)
                if m:
                    deps.append(Dependency(name=pkg, version=m.group(1)))
    return deps


# ---------------------------------------------------------------------------
# poetry.lock
# ---------------------------------------------------------------------------


def _parse_poetry_lock(path: Path) -> list[Dependency]:
    if tomllib is None:
        return []
    try:
        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return []
    deps: list[Dependency] = []
    for pkg in data.get("package", []):
        name = pkg.get("name")
        version = pkg.get("version")
        if name and version:
            deps.append(Dependency(name=name, version=str(version)))
    return deps


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

_CANDIDATES: list[tuple[str, str]] = [
    ("poetry.lock", "poetry.lock"),
    ("Pipfile.lock", "Pipfile.lock"),
    ("requirements.txt", "requirements.txt"),
    ("pyproject.toml", "pyproject.toml"),
]

_PARSERS = {
    "poetry.lock": _parse_poetry_lock,
    "Pipfile.lock": _parse_pipfile_lock,
    "requirements.txt": _parse_requirements_txt,
    "pyproject.toml": _parse_pyproject_toml,
}


def detect_dependencies(project_dir: Path) -> DetectionResult | None:
    """Find the highest-priority manifest in *project_dir* and parse its dependencies.

    Returns ``None`` if no supported manifest is found.
    """
    for filename, fmt in _CANDIDATES:
        candidate = project_dir / filename
        if candidate.is_file():
            deps = _PARSERS[fmt](candidate)
            return DetectionResult(
                manifest_path=str(candidate),
                manifest_format=fmt,
                dependencies=deps,
            )
    return None
